### PR TITLE
fix alt virtual host generation 

### PR DESCRIPTION
**Pull request by rshriram**
cc @ZackButcher @ijsnellf @GregHanson

Signed-off-by: Shriram Rajagopalan <shriramr@vmware.com>

### CODE REVIEW

**Comment by rshriram** [comment]
fixes #4845

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 1]
```diff
@@ -169,14 +169,37 @@ func (configgen *ConfigGeneratorImpl) BuildSidecarOutboundHTTPRouteConfig(env mo
...
+// function returns nil
+func generateAltVirtualHosts(hostname string, port int, proxyDomain string) []string {
 	var vhosts []string
-	for i := len(hostname) - 1; i >= 0; i-- {
-		if hostname[i] == '.' {
-			variant := hostname[:i]
+	uniqHostname, sharedDNSDomain := getUniqueAndSharedDNSDomain(hostname, proxyDomain)
+
+	if len(sharedDNSDomain) == 0 {
+		return nil // shared dns domain would be
```
Comment (ZackButcher) [inline pilot/pkg/networking/core/v1alpha3/httproute.go]: shared dns domain would be _ ?

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 2]
```diff
@@ -209,3 +232,56 @@ func mergeAllVirtualHosts(vHostPortMap map[int][]route.VirtualHost) []route.Virt
...
+// foo.ns2.svc.cluster.local). In this case, if the proxy is in ns2.svc.cluster.local, then while
+// generating alt virtual hosts for service foo.ns1 for the sidecars in ns2 namespace, we should generate
+// foo.ns1, foo.ns1.svc, foo.ns1.svc.cluster.local and should not generate a virtual host called "foo" for
+// foo.ns1 service.
+// So given foo.ns1.svc.cluster.local and ns2.svc.cluster.local, this function will return
+// foo.ns1, and svc.cluster.local.
+// When given foo.ns2.svc.cluster.local and ns2.svc.cluster.local, this function will return
+// foo, ns2.svc.cluster.local.
+func getUniqueAndSharedDNSDomain(fqdnHostname, proxyDomain string) (string, string) {
+
```
Comment (ZackButcher) [inline pilot/pkg/networking/core/v1alpha3/httproute.go]: super nit: no newline here

**Comment by ZackButcher** [comment]
+1 to Isaiah's comment.

**Comment by rshriram** [comment]
added tests.. PTAL

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 3]
```diff
@@ -168,6 +156,24 @@ func (configgen *ConfigGeneratorImpl) BuildSidecarOutboundHTTPRouteConfig(env mo
 	return out
 }
 
+// buildVirtualHostDomains generates the set of domain matches for a service being accessed from
+// a proxy node
+func buildVirtualHostDomains(service *model.Service, port int, node model.Proxy) []string {
+	domains := []string{service.Hostname, fmt.Sprintf("%s:%d", service.Hostname, port)}
+	if !service.MeshExternal {
```
Comment (ZackButcher) [inline pilot/pkg/networking/core/v1alpha3/httproute.go]: Add a comment like "for external services, don't generate alternate names but use only the user-provided name" or similar? (It's not immediately obvious why this check exists, so lets explain the business logic here for later reference.)

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 4]
```diff
@@ -168,6 +156,24 @@ func (configgen *ConfigGeneratorImpl) BuildSidecarOutboundHTTPRouteConfig(env mo
...
+		domains = append(domains, generateAltVirtualHosts(service.Hostname, port, node.Domain)...)
+	}
+	if len(service.Address) > 0 {
+		// add a vhost match for the IP (if its non CIDR)
+		cidr := util.ConvertAddressToCidr(service.Address)
```
Comment (ZackButcher) [inline pilot/pkg/networking/core/v1alpha3/httproute.go]: Ug, I did _not_ see that implemented our own CIDR parsing. I'd have stopped that dead in a PR. We should not do that ourselves. Can you please delete this method in the util and just use [`net.ParseCIDR`](https://golang.org/pkg/net/#ParseCIDR).

File: pilot/pkg/networking/core/v1alpha3/httproute_test.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+			node: model.Proxy{
+				Domain: "example.com",
+			},
+			want: []string{"foo.local.campus.net", "foo.local.campus.net:80"},
+		},
+	}
+
+	for _, c := range cases {
+		out := buildVirtualHostDomains(c.service, c.port, c.node)
+		if !checkEquals(out, c.want) {
```
Comment (ZackButcher) [inline pilot/pkg/networking/core/v1alpha3/httproute_test.go]: `reflect.DeepEquals(out, c.want)` and you can delete `checkEquals`

File: pilot/pkg/networking/core/v1alpha3/httproute.go (hunk 4 above)
Comment (ijsnellf) [reply pilot/pkg/networking/core/v1alpha3/httproute.go]: We should be careful. In validation we allow IP or CIDR for external services. `net.ParseCIDR` won't allow IPs.

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 6]
```diff
@@ -209,3 +240,55 @@ func mergeAllVirtualHosts(vHostPortMap map[int][]route.VirtualHost) []route.Virt
...
+func getUniqueAndSharedDNSDomain(fqdnHostname, proxyDomain string) (string, string) {
+	// split them by the dot and reverse the arrays, so that we can
+	// start collecting the shared bits of DNS suffix.
+	// E.g., foo.ns1.svc.cluster.local -> local,cluster,svc,ns1,foo
+	//       ns2.svc.cluster.local -> local,cluster,svc,ns2
+	partsFQDN := reverseArray(strings.Split(fqdnHostname, "."))
+	partsProxyDomain := reverseArray(strings.Split(proxyDomain, "."))
+	var sharedSuffixesInReverse []string // pieces shared between proxy and svc. e.g., local,cluster,svc
+	i := 0
+	for ; i < min(len(partsFQDN), len(partsProxyDomain)); i++ {
```
Comment (ijsnellf) [inline pilot/pkg/networking/core/v1alpha3/httproute.go]: nit: `for i := 0; i < ...; i++ {`

File: pilot/pkg/networking/core/v1alpha3/httproute.go (hunk 4 above)
Comment (ZackButcher) [reply pilot/pkg/networking/core/v1alpha3/httproute.go]: Ah, good call (IMO it's pretty bad that it doesn't just parse the IP as part of the /32 subnet). This isn't so bad given we're returning an Envoy config object anyway.

File: pilot/pkg/networking/core/v1alpha3/httproute.go (hunk 3 above)
Comment (rshriram) [reply pilot/pkg/networking/core/v1alpha3/httproute.go]: Actually I realized that now that we have this logic in place, we no longer need that if block

**Pull request merged**